### ISSUE TITLE
Add an option for returning in SDK types in search.getDeclerations

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#32446]
Recently a new API was added to the analysis server for searching declerations (#29510). During discussions there was a comment:

> We need to decide whether we want declarations from explicit or know files. Or it could be a flag on the request. But only if you actually going to use it :-)

We didn't add sch a flag, but the request has some up from someone testing the new search in a Dart Code beta:

https://github.com/Dart-Code/Dart-Code/issues/659

Is it reasonable simple to add? I could easily add a config option in Dart Code that we could pass through if so.

(cc @bwilkerson @scheglov)

**Comment by bwilkerson** [comment]
It shouldn't be too hard to add declarations from the SDK.

I'd recommend adding a single flag that will only include declarations in the SDK. If we find we want declarations from packages at a later point, then I would add a second flag for those.

**Comment by DanTup** [comment]
Assuming it doesn't add much work, is it worth adding both flags together to avoid coming back later? I suspect it'll be the next question (especially since I think there's a preference to add things in packages than to the SDK these days). If you support both, I'll definitely add options (I already have some options for SDK + Libs like debugging, so it wouldn't seem weird having two).

**Comment by bwilkerson** [comment]
I don't think it would add much in terms of work, but

(a) I hesitate to do work speculatively when there's so much work that we _know_ would be valuable and

(b) one of the reasons we don't currently search packages is because of the additional overhead of having to analyze all of the libraries in all of the packages, whether they are used or not, and that's still a concern.

I personally would prefer to have both the SDK and all available packages included in search results, used for completions, etc. So I'm in favor, conceptually, of doing this. But I've been convinced that it might be too expensive. @scheglov can probably present the argument against better than I.

**Comment by DanTup** [comment]
I think when I said "work" I meant "overhead" (work for the server!). If the data isn't all already there in memory then I understand the concern.

The original request came from @pulyaevskiy who said in Gitter "especially for Flutter as they have a lot of documentation in code". I'm not sure whether the code in question is SDK or libraries. I'll let him add his opinions.

We could always see what feedback we get as the new version rolls out and the SDK results disappear from the search. If we do add a flag for just SDK we can always collect stats on how many people are using it too.

**Comment by bwilkerson** [comment]
Good point. The Flutter code is in a package. (Other than the underlying support in `dart:ui`.)

**Comment by DanTup** [comment]
Btw, is the extra work required to scan the packages up-front (eg. you have to do it even if the user never sends this flag) or at the time you get the flag?

If it's the latter, then people who haven't opted-in to this wouldn't be affected. If it's the former then maybe there should be a way for a client to opt-in to indexing of it at startup?

**Comment by bwilkerson** [comment]
We could implement it either way.

As you noted, the disadvantage of doing it early is that users pay for it whether they use it or not. The disadvantage of waiting as late as possible is that the response might be very slow the first time the request is sent.

**Comment by DanTup** [comment]
There's always the option for me to send a "dummy" request (eg. a search with no criteria and a maxResults of 1) if the user has this enabled in order to begin the fetching. This could give us the best of both worlds without a new API to start the indexing? It doesn't seem like the worst hack in the world :-)

**Comment by pulyaevskiy** [comment]
>  I'm not sure whether the code in question is SDK or libraries. I'll let him add his opinions.

This is a good point. I was thinking about SDK code to be everything which comes with said SDK, so for Dart SDK it would be everything from `dart:*` and for Flutter SDK it would also add `package:flutter/*`. 

Though thinking more about it I probably was looking for a way to lookup any definition in project dependencies. Here is how UX for this feature works in Intellij:
![kapture 2018-03-07 at 9 56 52](https://user-images.githubusercontent.com/441756/37109187-10261358-21ee-11e8-8117-aa3a091c0024.gif)

Hot key is `Cmd+O`, hitting it again while the dialog is open toggles "Include non-project dependencies" checkbox, which I find really convenient.

Also there seems to be no performance penalty for searching in project dependencies in Intellij, wondering if they have local cache of all SDK/packages or something...

**Comment by DanTup** [comment]
Sadly VS Code doesn't let us have a toggle like that. The performance hit will come during initial indexing, so probably done at project startup rather than as you're typing.

How does that tickbox affect Flutter? Since `package:flutter/` is a package and not SDK (presumably), so is it included with or without the tickbox?

**Comment by bwilkerson** [comment]
> Also there seems to be no performance penalty for searching in project dependencies in Intellij, wondering if they have local cache of all SDK/packages or something...

Yes, IntelliJ indexes all of the files at start up. That's part of what makes it take a while before the IDE is completely ready to use. You can also see it right after switching branches. If you ask to open declarations it will tell you that you can't until it's done indexing (and then opens the open file dialog instead).

**Comment by pulyaevskiy** [comment]
> How does that tickbox affect Flutter? Since package:flutter/ is a package and not SDK (presumably), so is it included with or without the tickbox?

When it's off it works the same as in DartCode 2.10.0 beta (includes only types declared in the project, no flutter, dart or other dependencies). When it's on it includes everything (`dart:*`, `package:*`.

**Comment by scheglov** [comment]
FYI, when I was talking about adding "known files", I thought about files that are not part of analysis roots (i.e. code that is in user projects), but still implicitly referenced, because it is part of code that is transitively (and only partially) analyzed - in SDK or packages. This does not include all potential corpus of SDK and packages code, e.g. code that is not referenced at all.

There are at least two other issues that could be solved by indexing all potential code, even not yet referenced:
1. Quick Fix for imports. Currently, when you add a new package, and try to reference a class defined in it, you don't get a fix, because we parse files only when they are referenced. But in this case they are not - we've just imported it!

2. Type hierarchies for not yet imported libraries. This is again something IntelliJ users often expect to have, but we don't help them. Currently type hierarchies require full file resolution, because they are built on top of the general purpose, resolution based index. In theory we could reimplement some parts of resolution for resolving only names in class extends/implements/with clauses. To do this we need only unlinked summaries (just parsing, cacheable).

Any analysis of "accessible" files, which are neither added, nor implicitly referenced, should be on demand. Cacheable, but demand. We don't want other users, such as analyzer_cli, to index the whole world, when all they want is errors.

**Comment by DanTup** [comment]
I don't know enough to have an opinion on much of the above so I'll leave it to you two :-)

> Any analysis of "accessible" files, which are neither added, nor implicitly referenced, should be on demand. Cacheable, but demand. We don't want other users, such as analyzer_cli, to index the whole world, when all they want is errors.

This makes sense. And if there's a way for an editor to start the process (whether by a specific API or just sending a dummy search) we can do that based on the user setting (so we only do it for users that want this behaviour).

**Comment by DanTup** [comment]
For anyone using this in VS Code, you should upgrade to v2.14 Beta 1 as there's a serious performance issue when VS Code gets a large number of results here. In v2.14 we'll truncate to first 500 - we may revert this once there's a response to/fix for https://github.com/Microsoft/vscode/issues/51236
